### ISSUE TITLE
DQN is broken

### ISSUE THREAD

**Issue by zspasztori** [pytorch/tutorials#263]
Hi,

I am running your implementation of DQN.
Tried several runs, but it does not seem to converge with the given hyperparameters.
Tried higher amount of episodes in the magnitude of 1000, but there is no convergence.

Also the ploting has problems, you need to move the display part before sleep.
![dqn_tutorial](https://user-images.githubusercontent.com/12113563/41774606-b7066c82-7620-11e8-938e-87b726209518.png)

cc @vmoens @nairbv

**Comment by johan-gras** [comment]
![figure_2](https://user-images.githubusercontent.com/30185046/43036611-58247f14-8d05-11e8-850a-126350f514b1.png)

Same issue here.
Running the exact code from the tutorial for 10 000 episodes and the algorithm doesn't converge.
I've been testing lots of differents hyperparameters without success.

**Comment by cebenso2** [comment]
I am experiencing the same issue.

**Comment by leonardblier** [comment]
I also have the same issue.

**Comment by danaugrs** [comment]
Same:

![pytorch-dqn-example](https://user-images.githubusercontent.com/5261448/51807454-1e344600-226e-11e9-8f3f-e066ec3b255d.png)

**Comment by vmikulik** [comment]
+1 to the plotting code not working (on MacOS, using Jupyter). Swapping display and clear_output solved the issue for me.

**Comment by mansur007** [comment]
Has anyone solved this problem?
I am having it too...

**Comment by zhihaocheng** [comment]
same too

**Comment by mansur007** [comment]
> Has anyone solved this problem?
> I am having it too...

If you anyone is still looking for a working DQN - check out mine: https://github.com/mansur007/gym_cartpole_dqn

**Comment by josiahls** [comment]
After making my own DQN. My main concern:
- Why is the target network being updated every few **episodes**. What DQN paper is that supposed to be related to? Surely you should be doing per step, like every 5-10 steps update the target network.

As noted:
![image](https://user-images.githubusercontent.com/19930483/65838095-76c05500-e2cd-11e9-8327-21fcf49437d5.png)
[1] Mnih, Volodymyr, et al. "Playing atari with deep reinforcement learning."
            arXiv preprint arXiv:1312.5602 (2013).

Key quote "parameters from the previous iteration θi−1 are held fixed when optimising the loss function" they do not say previous episode, they say previous iteration.

**I recommend changing your target updates to every few iterations.**

This is because since the target network is be initialized randomly, it will need a chance to correct itself / iron out pockets of incorrent Q values. Every few episodes seems way too slow and I fould it takes hundreds / thousands of target updates to get a decent target network.

I recommend for those learning this to experiment with the gym_maze envs _without exploration_ to make sure your DQN is learning properly. Cartpole is way to hard to debug your DQN when starting out.

**Comment by yizhixiaotaozi** [comment]
Same problem.  Adjusting parameters is a HARD work  T-T .

**Comment by thomashirtz** [comment]
Did someone manage to make it work ? Is it possible to share the parameter used ? I really cannot find any set of parameters that work

**Comment by StoyanVenDimitrov** [comment]
Any improvements here?

**Comment by mansur007** [comment]
I remember making things work here https://github.com/mansur007/gym_cartpole_dqn/ but the code is from ~2 years ago...

**Comment by Erlix322** [comment]
Sorry to bother, but Any updates?

**Comment by carljparker** [comment]
I believe that this issue is resolved by the work done for #2030.
